gpt_image/partition.py

```diff
@@ -175,46 +240,98 @@ def marshal(self) -> bytes:
         partition_bytes = self._PARTITION_FORMAT.pack(
             uuid.UUID(self.type_guid).bytes_le,
             uuid.UUID(self.partition_guid).bytes_le,
-            self.first_lba,
-            self.last_lba,
+            self.first_lba_staged,
+            self.last_lba_staged,
             attributes_value,
             bytes(self.partition_name, encoding="utf_16_le"),
         )
         return partition_bytes
 
-    def write_data(self, disk: Disk, data: bytes) -> int:
+    def _staged_attrs(self):
+        return [ attr
+            for attr in [self._first_lba, self._last_lba, self._size]
+            if attr.needs_commit()
+        ]
+
+    def _commit_attrs(self):
+        for attr in self._staged_attrs():
+            attr.commit()
+
+    def needs_commit(self):
+        return len(self._staged_attrs()) != 0
+
+    def commit(self, disk: Disk, tmpfile: Optional[IO[bytes]] = None):
+        if tmpfile is None:
+            f = open(disk.image_path, 'r+b')
+        else:
+            f = tmpfile
+        try:
+            max_size = min(self.size, self.size_staged)
+            data = self.read(disk, max_size=max_size)
```

Comment:
this section would probably benefit from some chunking (to avoid loading the entire partition into memory at once)

Reply:
Agreed. This isn't a big issue for my use-case (the images are very small), so it can be addressed in a follow up.